### PR TITLE
Feature language c support

### PR DESCRIPTION
**Pull request by fedescarpa**

### CODE REVIEW

**Comment by fedescarpa** [comment]
Documentation http://hackage.haskell.org/package/language-c-0.8.3/docs/Language-C-Parser.html

**Comment by fedescarpa** [comment]
### Example 1

```c
int a, b = 0, *c, *d = 1, e[], f[3], g[4] = {1, 2, 3, 4}, h = i(3);
```

```haskell
Right
  (
    CTranslUnit
      [
        CDeclExt
          (
            CDecl
              [
                CTypeSpec (CIntType _)
              ]
              [
                (
                  Just (CDeclr (Just (Ident "a"  97 _)) []                                                           Nothing   []   _),
                  Nothing,
                  Nothing
                ),
                (
                  Just (CDeclr (Just (Ident "b"  98 _)) []                                                           Nothing   []   _),
                  Just (CInitExpr (CConst (CIntConst 0 _)) _),
                  Nothing
                ),
                (
                  Just (CDeclr (Just (Ident "c"  99 _)) [CPtrDeclr [] _]                                             Nothing   []   _),
                  Nothing,
                  Nothing
                ),
                (
                  Just (CDeclr (Just (Ident "d" 100 _)) [CPtrDeclr [] _]                                             Nothing   []   _),
                  Just (CInitExpr (CConst (CIntConst 1 _)) _),
                  Nothing
                ),
                (
                  Just (CDeclr (Just (Ident "e" 101 _)) [CArrDeclr [] (CNoArrSize False) _]                          Nothing   []   _),
                  Nothing,
                  Nothing
                ),
                (
                  Just (CDeclr (Just (Ident "f" 102 _)) [CArrDeclr [] (CArrSize False (CConst (CIntConst 3 _))) _]   Nothing   []   _),
                  Nothing,
                  Nothing
                ),
                (
                  Just (CDeclr (Just (Ident "g" 103 _)) [CArrDeclr [] (CArrSize False (CConst (CIntConst 4 _))) _]   Nothing   []   _),
                  Just (CInitList [([],CInitExpr (CConst (CIntConst 1 _)) _),([],CInitExpr (CConst (CIntConst 2 _)) _),([],CInitExpr (CConst (CIntConst 3 _)) _),([],CInitExpr (CConst (CIntConst 4 _)) _)] _),
                  Nothing
                ),
                (
                  Just (CDeclr (Just (Ident "h" 104 _)) [] Nothing [] _),
                  Just (CInitExpr (CCall (CVar (Ident "i" 105 _) _) [CConst (CIntConst 3 _)] _) _),
                  Nothing
                )
              ]
              _
          )
      ]
      _

  )
```
### Example 2

```c
typedef struct persona { 
  char* name; 
  int age; 
} t_persona;
```

```haskell
Right
  (
    CTranslUnit
      [
        CDeclExt
          (
            CDecl
              [
                CStorageSpec (CTypedef _),
                CTypeSpec (CSUType (CStruct CStructTag (Just (Ident "persona" 244656735 _)) (Just [CDecl [CTypeSpec (CCharType _)] [(Just (CDeclr (Just (Ident "name" 213610734 _)) [CPtrDeclr [] _] Nothing [] _),Nothing,Nothing)] _,CDecl [CTypeSpec (CIntType _)] [(Just (CDeclr (Just (Ident "age" 1668065 _)) [] Nothing [] _),Nothing,Nothing)] _]) [] _) _)
              ]
              [
                (
                  Just (CDeclr (Just (Ident "t_persona" 446179911 _)) [] Nothing [] _),
                  Nothing,
                  Nothing
                )
              ]
            _
          )
      ]
      _
  )
```

### Example 3

```c
typedef struct persona {
  char* name;
  int age;
} t_persona;

int edad(t_persona p) {
  return p.edad;
}
```

```haskell
Right
  (
    CTranslUnit
      [
        CDeclExt
          (
            CDecl
              [
                CStorageSpec (CTypedef _),
                CTypeSpec
                  (
                    CSUType
                      (
                        CStruct
                          CStructTag
                          (
                            Just (Ident "persona" 244656735 _)
                          )
                          (
                            Just [
                              CDecl [CTypeSpec (CCharType _)] [(Just (CDeclr (Just (Ident "name" 213610734 _)) [CPtrDeclr [] _] Nothing [] _), Nothing, Nothing)] _,
                              CDecl [CTypeSpec (CIntType _ )] [(Just (CDeclr (Just (Ident "age"  1668065   _)) []               Nothing [] _), Nothing, Nothing)] _
                            ]
                          )
                          []
                          _
                      )
                      _
                  )
              ]
              [
                (Just (CDeclr (Just (Ident "t_persona" 446179911 _)) [] Nothing [] _),  Nothing,  Nothing)
              ]
              _
          ),

        CFDefExt
          (
            CFunDef
              [
                CTypeSpec (CIntType _)
              ]
              (
                CDeclr
                  (
                    Just (Ident "edad" 211317349 _)
                  )
                  [
                    CFunDeclr
                      (
                        Right
                          (
                            [
                              CDecl
                                [
                                  CTypeSpec (CTypeDef (Ident "t_persona" 446179911 _) _)
                                ]
                                [
                                  (Just (CDeclr (Just (Ident "p" 112 _)) [] Nothing [] _),  Nothing,  Nothing)
                                ]
                                _
                            ],
                            False
                          )
                      )
                      []
                      _
                  ]
                  Nothing
                  []
                  _
              )
              []
              (
                CCompound [] [CBlockStmt (CReturn (Just (CMember (CVar (Ident "p" 112 _) _) (Ident "edad" 211317349 _) False _)) _)] _
              )
              _
          )
      ]
      _
  )
```

**Comment by flbulgarelli** [comment]
Wooooowwwwwww :clap: :clap: 

This is awesome :heart:

**Comment by julian-berbel** [comment]
:heart:

File: src/Language/Mulang/Parsers/C.hs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+muStructureUnionTag (CStruct structTag _ _ _ _) = muStructTag structTag
+
+muStructTag CStructTag = "struct"
+muStructTag CUnionTag  = "union"
+
+muVarInit e = debug e
+
+-- -- Combinators
+
+fmapOrNone f = fromMaybe None . fmap f
```
Comment (julian-berbel) [inline src/Language/Mulang/Parsers/C.hs:85]: This can also be written as
`fmapOrNone = maybe None`
:grin:

**Comment by luchotc** [comment]
:tada:

**Comment by fedescarpa** [comment]
With @julian-berbel we have finished support for the language. Support for enums, structs, and typedefs is still missing but we think it's ready to be merge into master and add that support later.

File: spec/CSpec.hs [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+      it "parses logical operators" $ do
+        run [text|
+          int main () {
+            a || b;
+          }
+          |] `shouldBe` cContext (Application (Primitive Or) [Reference "a", Reference "b"])
+
+      it "parses simple assignment" $ do
+        run [text|
```
Comment (flbulgarelli) [inline spec/CSpec.hs:138]: Could you please a test for unsupported features?

**Comment by julian-berbel** [comment]
Also see #275.

**Comment by julian-berbel** [comment]
We just got done with your suggestions with @fedescarpa!

**Pull request merged**